backend/apps/beaver_api/beaver_api/code_api/serializers.py

```diff
@@ -1,19 +1,55 @@
 from code_api.models import CodeDocument
 
+from django.core.exceptions import ValidationError
+from django.core.validators import URLValidator
+from django.utils.translation import gettext_lazy as _
+from language_api.models import Language
 from rest_framework import serializers
+from tags_api.models import Tag
 
 
-class CodeDocumentSerializer(serializers.HyperlinkedModelSerializer):
+class CodeDocumentSerializer(serializers.ModelSerializer):
+    id = serializers.UUIDField(required=False)
+    tags = serializers.SlugRelatedField(
+        many=True,
+        slug_field="name",
+        queryset=Tag.objects.all(),
+        error_messages={
+            "does_not_exist": "Tag with name {value} does not exist."
+        },
+    )
+    language = serializers.SlugRelatedField(
+        slug_field="name",
+        queryset=Language.objects.all(),
+        error_messages={
+            "does_not_exist": "Language with name {value} does not exist."
+        },
+    )
+
     class Meta:
         model = CodeDocument
-        fields = (
-            "id",
-            "title",
-            "code",
-            "link_to_project",
-            "language",
-            "tags",
-            "created_at",
-            "updated_at",
-        )
+        fields = "__all__"
         read_only_fields = ("id", "created_at", "updated_at")
+
+    def validate_title(self, value):
+        if CodeDocument.objects.filter(title=value).exists():
+            raise serializers.ValidationError("Title must be unique.")
+        return value
+
+    def validate_code(self, value):
+        if CodeDocument.objects.filter(code=value).exists():
+            raise serializers.ValidationError(
+                "Code snippet already exists in the database."
+            )
+        return value
```

Comment:
Plan is to overwrite standard validation method and check if record with same title and code snippet exists in database. If yes - raise error. What do you think?

Reply:
I think it's better to create a method field on the serializer for the code hash and move there the logic from your 'BaseCodeDocumentView'.  The hash field of the model already forces the uniqueness of the code (if hash computation works correctly), I think it is a stronger check.  We also applied a unique constraint on the code itself, so that should work as well. Could you please check how Django creates enqueue constraint in the db, we do not want pg uses the whole text it is not efficient, we that pg uses some hash to guarantee uniqueness (see example https://stackoverflow.com/a/67715589). 

I am not sure, about filtering by code, I think it is quite slow, so it is better to use a hash. But you can always check what Django does for that query and if Django does something clever here, you can leave it like that :). (use 'python manage.py shell_plus --print-sql'  to see SQL queries )